### ISSUE TITLE
Possibility to modify height of inline viewport

### ISSUE THREAD

**Issue by PitiBouchon** [ratatui-org/ratatui#984]
I would like to use the `Viewport::Inline(height)` and modify the `height` of it

I'm looking for :
- Display one or more lines of modifiable text (e.g. `docker compose up`)
- The number of lines of text will grow over time
- `ViewPort::Inline` doesn't support the height changing

`insert_before` already works to print a line before the viewport but isn't modifiable

**Comment by joshka** [comment]
This seems like a reasonable feature to add. Would you mind putting some thoughts down on how the library might change to accommodate this (i.e. what methods will be added to the relevant types)?

**Comment by PitiBouchon** [comment]
I currently don't have much free time, but I'll look into it in the future

**Comment by joshka** [comment]
> I currently don't have much free time, but I'll look into it in the future

No problem. Take as much time as you need.

**Comment by nfachan** [comment]
This is something that we would like to use for our [Maelstrom project](https://github.com/maelstrom-software/maelstrom). As such, I have a few thoughts on it. Unfortunately, there seem to be a lot of details to conside#

## Terminal Capabilities
The [terminfo manpage](https://man7.org/linux/man-pages/man5/terminfo.5.html) has some interesting content in the "   Insert/delete line and vertical motions" section that is relevant to this discussion.

I think if done right, and on a modern terminal, it would be possible to do viewport resizing without much redrawing, and thus without a lot of flickering. Unfortunately, this complicates the API because we'd need to give some hints to Ratatui so that it would scroll/insert/delete in the right place to minimize redrawing.

## Growing the Inline Viewport

When growing the inline viewport, there are a few different issues to consider. The first is whether or not the viewport currently sits at the bottom of the screen or not. If it does, then it will obviously have to grow upwards. But if it doesn't, then, in theory, it could grow upwards or downwards.

For the sake of simplicity, I'm going to that the viewport is only growing by a single line, because things can be complicated if the viewport isn't sitting at the bottom of the screen, it's growing downwards, but there's not enough room to accommodate the whole operation. We can always break down a multi-line growth into a series of single-line grows.

If the viewport is growing up, then there is another question of what to do with the lines currently above the viewport. Should we overwrite them, or should we scroll them up?

Finally, in order to minimize redrawing, we want to know where in the viewport the new line is going to appear: top, bottom, or middle.

With all of that, you get something like (in addition to updating the `viewport_area`):

Scroll or Overwrite | Up or Down | On Bottom? | Where in Viewport | Operation
-|-|-|-|-
Scroll | Up | Yes | Anywhere | Create a scroll region from top of screen to desired line and scroll up or use terminal's insert line.
Scroll | Up | No | Anywhere | Weird case. May not support. Create a scroll region from top of screen to desired line and scroll up or possibly use terminal's insert line.
 - | Down | No | Anywhere | Create a scroll region from desired line to one below current viewport and scroll down or possibly use terminal's insert line.
Overwrite | Up | Yes or No | Top | Clear the line just added to the viewport.
Overwrite | Up | Yes or No | Not Top | Either create a scroll region inside of the viewport to scroll up the currently-written contents, or accept redrawing the top part of the viewport.
Overwrite | Down | Yes or No | Bottom | Clear the line just added to the viewport.
Overwrite | Down | Yes or No | Not Bottom | Either create a scroll region inside of the viewport to scroll down the currently-written contents, or accept redrawing the bottom part of the viewport.

**Comment by joshka** [comment]
I'd recommend experimenting how each of these works outside of ratatui (I.e using plain crossterm), with some sample code and then mapping the ideas onto ratatui. Note how resizing the double buffer interacts here. Most of the time we're just redrawing the entire buffer when there's a resize, as accurately determining the diff is difficult (this seems like an example of the sort of difficulty to do that). Also check how this interacts with a few terminal emulators if you have a way to do so (Mac: iterm2 / terminal.app, windows: ?, Linux: ?, cross platform: alacrity, kitty, wezterm).

It would be easy to end up writing this in an unmaintainable mess of special cases, but simple, clean and obvious would make me happy to see this. With an understanding of what happens when terminal support is lacking would be nice to have too.

**Comment by domenkozar** [comment]
I've looked into [Ink](https://github.com/vadimdemedes/ink) and they handle this by having a virtually calculated height so they can translate that into terminal codes.

If I understand correctly for Ratatui one would have to render twice, once to get the height and then second time with the resize.

EDIT: looks like https://github.com/joshka/tui-brulee

**Comment by joshka** [comment]
> If I understand correctly for Ratatui one would have to render twice, once to get the height and then second time with the resize.

If you want something super robust, take a look at some of the ideas that masory (part of xilem) uses for their GUI stuff. They have a multi-pass approach (layout, update, etc.). https://docs.rs/masonry_winit/latest/masonry_winit/doc/index.html has some info. Also bevy has a different / similar idea in the way they do things with their ECS approach.

Also, there's a bunch of stuff happening on this in https://github.com/ratatui/ratatui/pull/1964 right now. I'm working through some of the deeper parts to make sure that the existing code makes more sense as well as solving this correctly.

**Comment by domenkozar** [comment]
> > If I understand correctly for Ratatui one would have to render twice, once to get the height and then second time with the resize.
> 
> If you want something super robust, take a look at some of the ideas that masory (part of xilem) uses for their GUI stuff. They have a multi-pass approach (layout, update, etc.). https://docs.rs/masonry_winit/latest/masonry_winit/doc/index.html has some info. Also bevy has a different / similar idea in the way they do things with their ECS approach.
> 
> Also, there's a bunch of stuff happening on this in [#1964](https://github.com/ratatui/ratatui/pull/1964) right now. I'm working through some of the deeper parts to make sure that the existing code makes more sense as well as solving this correctly.

I've looked into Masory, but it doesn't seem to have a terminal render? I've found https://github.com/ccbrown/iocraft and I'm giving it a try.

**Comment by joshka** [comment]
Yup - to be clear, I'm talking about the concepts rather than fully working end to end code for masonry. I suspect it might be fairly easy to build a TUI framework on Masonry, but haven't looked deeply on it. Several people have mentioned it as an idea in the past however (on the Xilem Zulip).
